src/sysupdate/updatectl.c

```diff
@@ -0,0 +1,1226 @@
+                        JSON_VARIANT_ARRAY_FOREACH(entry, value) {
+                                assert(sd_json_variant_is_string(entry));
+                                r = strv_extend(&v.changelog, sd_json_variant_string(entry));
+                                if (r < 0)
+                                        log_error_errno(r, "Failed to parse JSON: %m");
+                        }
+                } else if (streq(key, "contents")) {
+                        r = sd_json_variant_format(value, 0, &v.contents_json);
+                        if (r < 0)
+                                log_error_errno(r, "Failed to parse JSON: %m");
+                } else
+                        assert_not_reached();
+        }
+
+        *ret = TAKE_STRUCT(v);
+        return 0;
+}
+
+static int list_versions_finished(sd_bus_message *reply, void *userdata, sd_bus_error *ret_error) {
+        _cleanup_(userdata_freep) Userdata *data = ASSERT_PTR(userdata);
+        Table *table = ASSERT_PTR(data->userdata);
+        const sd_bus_error *e;
+        _cleanup_(version_clear) Version v = {};
+        _cleanup_free_ char *version_link = NULL;
+        const char *color;
+        int r;
+
+        assert(reply);
+
+        e = sd_bus_message_get_error(reply);
+        if (e)
+                return log_bus_error(0, e, NULL, "call Describe");
+
+        r = parse_describe(reply, &v);
+        if (r < 0)
+                return r;
+
+        color = update_set_flags_to_color(v.flags);
+
+        if (urlify_enabled() && v.changelog) {
+                version_link = strjoin(v.version, special_glyph(SPECIAL_GLYPH_EXTERNAL_LINK));
+                if (!version_link)
+                        return log_oom();
+        }
+
+        r = table_add_many(table,
+                           TABLE_STRING,    update_set_flags_to_glyph(v.flags),
+                           TABLE_SET_COLOR, color,
+                           TABLE_STRING,    version_link ?: v.version,
+                           TABLE_SET_COLOR, color,
+                           TABLE_SET_URL,   v.changelog,
+                           TABLE_STRING,    update_set_flags_to_string(v.flags),
+                           TABLE_SET_COLOR, color);
+        if (r < 0)
+                return table_log_add_error(r);
+
+        return 0;
+}
+
+static int list_versions(sd_bus *bus, const char *target_path) {
+        _cleanup_(sd_event_unrefp) sd_event *event = NULL;
+        _cleanup_(table_unrefp) Table *table = NULL;
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL;
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_strv_free_ char **versions = NULL;
+        unsigned remaining = 0;
+        int r;
+
+        r = sd_bus_call_method(bus, bus_sysupdate_mgr->destination, target_path, SYSUPDATE_TARGET_INTERFACE, "List", &error, &reply, "b", arg_offline);
+        if (r < 0)
+                return log_bus_error(r, &error, NULL, "call List");
```

Comment:
better error messages please

Reply:
The log_bus_error wrapper expands this to `Failed to call List for target '<target>': <bus error message>`, which should be fine, I hope.